pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainGist failed

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainGist [failed](https://tanzanite.cluster.engflow.com/invocations/default/7a043f45-0327-499c-88c8-5a3f68932ec2?testReportRun=12&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on master @ [688e82e8d015350fe3aa263484416d28b232a25d](https://github.com/cockroachdb/cockroach/commits/688e82e8d015350fe3aa263484416d28b232a25d):

Fatal error:

```
panic: invalid datum type given: RECORD, expected RECORD [recovered]
	panic: invalid datum type given: RECORD, expected RECORD [recovered]
	panic: invalid datum type given: RECORD, expected RECORD [recovered]
	panic: invalid datum type given: RECORD, expected RECORD [recovered]
	panic: invalid datum type given: RECORD, expected RECORD [recovered]
	panic: invalid datum type given: RECORD, expected RECORD [recovered]
	panic: invalid datum type given: RECORD, expected RECORD

```
Stack: 

```
goroutine 2319 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0066d5a88, {0x892e758, 0xc006742c60}, {0x6954060?, 0xc0014c8490?})
	pkg/sql/conn_executor.go:1255 +0x19d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x892e758, 0xc006742c60}, {0xc006383bc8?})
	pkg/sql/conn_executor.go:962 +0x45
panic({0x6954060?, 0xc0014c8490?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	pkg/sql/colexecerror/error.go:137 +0x554
panic({0x6954060?, 0xc0014c8490?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	pkg/sql/colexecerror/error.go:137 +0x554
panic({0x6954060?, 0xc0014c8490?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	pkg/sql/colexecerror/error.go:137 +0x554
panic({0x6954060?, 0xc0014c8490?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	pkg/sql/colexecerror/error.go:137 +0x554
panic({0x6954060?, 0xc0014c8490?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	pkg/sql/colexecerror/error.go:137 +0x554
panic({0x6954060?, 0xc0014c8490?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/rowenc.DatumToEncDatum(...)
	pkg/sql/rowenc/encoded_datum.go:186
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc005339508)
	pkg/sql/plan_node_to_row_source.go:219 +0x2ed
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Next(0xc00a449508)
	pkg/sql/colexec/columnarizer.go:234 +0x11f
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next(0xc008d3c840)
	pkg/sql/colexec/invariants_checker.go:88 +0x42
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next(...)
	pkg/sql/colflow/stats.go:113
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc008d3c880?)
	pkg/sql/colexecerror/error.go:162 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next(0xc00b37b100)
	pkg/sql/colflow/stats.go:121 +0x4c
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).next(0xc0081d9ba0)
	pkg/sql/colexec/materializer.go:246 +0x6d
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).nextAdapter(...)
	pkg/sql/colexec/materializer.go:271
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0x10?)
	pkg/sql/colexecerror/error.go:162 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next(0xc0081d9ba0)
	pkg/sql/colexec/materializer.go:277 +0x46
github.com/cockroachdb/cockroach/pkg/sql.(*rowSourceToPlanNode).Next(0xc00b528a88, {{0x1?, 0x89a2ae0?}, 0x89a2ae0?, 0x1?})
	pkg/sql/row_source_to_plan_node.go:74 +0x3f
github.com/cockroachdb/cockroach/pkg/sql.(*bufferNode).Next(0xc004034b00, {{0x892e758?, 0xc003d90cc0?}, 0xc00b527008?, 0xc0066d6138?})
	pkg/sql/buffer.go:43 +0x83
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc002f0c308)
	pkg/sql/plan_node_to_row_source.go:211 +0xa8
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Next(0xc00a44a308)
	pkg/sql/colexec/columnarizer.go:234 +0x11f
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next(0xc008d3c980)
	pkg/sql/colexec/invariants_checker.go:88 +0x42
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next(...)
	pkg/sql/colflow/stats.go:113
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc008d3ca40?)
	pkg/sql/colexecerror/error.go:162 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next(0xc00b37b200)
	pkg/sql/colflow/stats.go:121 +0x4c
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).next(0xc00b1d7860)
	pkg/sql/colexec/materializer.go:246 +0x6d
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).nextAdapter(...)
	pkg/sql/colexec/materializer.go:271
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc003d90a20?)
	pkg/sql/colexecerror/error.go:162 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next(0xc00b1d7860)
	pkg/sql/colexec/materializer.go:277 +0x46
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).next(0xc005de3b30)
	pkg/sql/colflow/flow_coordinator.go:119 +0x35
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).nextAdapter(...)
	pkg/sql/colflow/flow_coordinator.go:136
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc0063809e0?)
	pkg/sql/colexecerror/error.go:162 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Next(0xc005de3b30)
	pkg/sql/colflow/flow_coordinator.go:141 +0x38
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run({0x892e758, 0xc003d90c30}, {0x895f020, 0xc005de3b30}, {0x88eef60, 0xc00278e808})
	pkg/sql/execinfra/base.go:193 +0x45
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc005de3b30, {0x892e720?, 0xc001d706e0?}, {0x88eef60, 0xc00278e808})
	pkg/sql/execinfra/processorsbase.go:727 +0x6d
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc00b51aa08, {0x892e720, 0xc001d706e0}, 0xc0?)
	pkg/sql/flowinfra/flow.go:574 +0x22c
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x89b51e0?, {0x892e720?, 0xc001d706e0?}, 0x80?)
	pkg/sql/colflow/vectorized_flow.go:300 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc007901600, {0x892e758, 0xc003d49d10}, 0xc0085f9380, 0xc008feb180, 0xc004035b80, 0xc00278e808, 0xc00b526008, 0xc006381890)
	pkg/sql/distsql_running.go:924 +0xd07
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).planAndRunSubquery(0xc007901600, {0x892e758, 0xc003d49d10}, 0x1, {{0x88d25a0, 0xc001cde050}, 0x3, 0x1, 0x0, {{0x892fb08, ...}, ...}, ...}, ...)
	pkg/sql/distsql_running.go:1876 +0x72d
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunSubqueries(0xc007901600, {0x892e758, 0xc003d49d10}, 0xc0066d6138, 0xc006381b48, {0xc008feb720, 0x2, 0x2}, 0xc0028a9008, 0xc006381ac0, ...)
	pkg/sql/distsql_running.go:1791 +0x1ba
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc007901600, {0x892e758, 0xc003d49d10}, 0xc0066d6640, 0xc0085f91e0, 0xc0066d6138, 0xc0028a9008, 0xc003d49da0)
	pkg/sql/distsql_running.go:1712 +0x1ae
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc0066d5a88, {0x892e758, 0xc003d49d10}, 0xc0066d6138, 0x3?, {0x7fe92b6abd38?, 0xc002f0e760?}, 0x0, 0xc00b3eb898, {0x88d25e0, ...})
	pkg/sql/conn_executor_exec.go:2435 +0x48a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0066d5a88, {0x892e758, 0xc003ca9110}, 0xc0066d6138, {0x7fe92b6abd38, 0xc002f0e760})
	pkg/sql/conn_executor_exec.go:1981 +0x119e
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0066d5a88, {0x892e758, 0xc003ca8ff0}, {{0x8954960, 0xc001cde3c0}, {0x0, 0x0, 0x0}, {0xc00b275000, 0x3ef}, ...}, ...)
	pkg/sql/conn_executor_exec.go:1187 +0x4a85
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x892e758?, 0xc003ca8ff0?})
	pkg/sql/conn_executor_exec.go:141 +0xb0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x892e758?, {0x892e758?, 0xc003ca8ff0?}, {0x8954960?, 0xc001cde3c0?}, 0xc00b275000?, 0x6?)
	pkg/sql/conn_executor_exec.go:3451 +0x2ec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0066d5a88, {0x892e758, 0xc003ca8ff0}, {{0x8954960, 0xc001cde3c0}, {0x0, 0x0, 0x0}, {0xc00b275000, 0x3ef}, ...}, ...)
	pkg/sql/conn_executor_exec.go:140 +0x725
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc00842f808, 0xc00842f8b8, 0xc0066d5a88, 0xc00842f4e0, 0xc006383940, 0xc006383930, 0xc006383960, 0xc00842f978)
	pkg/sql/conn_executor.go:2348 +0x3f1
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc0066d5a88)
	pkg/sql/conn_executor.go:2353 +0xa39
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0066d5a88, {0x892e758, 0xc006742c60}, 0x3c?, 0xc00193dd00?, 0xc0059811a0?)
	pkg/sql/conn_executor.go:2270 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc0069b5680?, {0x892e758?, 0xc006742c60?}, {0xc00469e000?}, 0x2?, 0x0?)
	pkg/sql/conn_executor.go:964 +0xda
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands(0xc002f0e008, {0x892e758, 0xc006742c60}, {0x0, 0x4, {0x1, {0x701bc9d, 0x3}, {0xc00757b3f0, 0xf}, ...}, ...}, ...)
	pkg/sql/pgwire/conn.go:251 +0x4c8
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func4()
	pkg/sql/pgwire/server.go:1196 +0xec
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl in goroutine 2317
	pkg/sql/pgwire/server.go:1193 +0x5db
```
<details><summary>Log preceding fatal error</summary>
<p>

```
*   | 	pkg/sql/colexec/materializer.go:246
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).nextAdapter
*   | 	pkg/sql/colexec/materializer.go:271
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	pkg/sql/colexecerror/error.go:162
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next
*   | 	pkg/sql/colexec/materializer.go:277
*   | github.com/cockroachdb/cockroach/pkg/sql.(*rowSourceToPlanNode).Next
*   | 	pkg/sql/row_source_to_plan_node.go:74
*   | github.com/cockroachdb/cockroach/pkg/sql.(*bufferNode).Next
*   | 	pkg/sql/buffer.go:43
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next
*   | 	pkg/sql/plan_node_to_row_source.go:211
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Next
*   | 	pkg/sql/colexec/columnarizer.go:234
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
*   | 	pkg/sql/colexec/invariants_checker.go:88
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
*   | 	pkg/sql/colflow/stats.go:113
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	pkg/sql/colexecerror/error.go:162
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
*   | 	pkg/sql/colflow/stats.go:121
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).next
*   | 	pkg/sql/colexec/materializer.go:246
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).nextAdapter
*   | 	pkg/sql/colexec/materializer.go:271
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	pkg/sql/colexecerror/error.go:162
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next
*   | 	pkg/sql/colexec/materializer.go:277
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).next
*   | 	pkg/sql/colflow/flow_coordinator.go:119
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).nextAdapter
*   | 	pkg/sql/colflow/flow_coordinator.go:136
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	pkg/sql/colexecerror/error.go:162
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Next
*   | 	pkg/sql/colflow/flow_coordinator.go:141
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
*   | 	pkg/sql/execinfra/base.go:193
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
*   | 	pkg/sql/execinfra/processorsbase.go:727
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	pkg/sql/flowinfra/flow.go:574
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run
*   | 	pkg/sql/colflow/vectorized_flow.go:300
* Wraps: (6) invalid datum type given: RECORD, expected RECORD
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) *assert.withAssertionFailure (5) *withstack.withStack (6) *errutil.leafError
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=12</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainGist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-44226

--- a/pkg/ccl/testccl/sqlccl/BUILD.bazel
+++ b/pkg/ccl/testccl/sqlccl/BUILD.bazel
@@ -42,6 +42,7 @@ go_test(
         "//pkg/sql",
         "//pkg/sql/catalog/lease",
         "//pkg/sql/catalog/replication",
+        "//pkg/sql/colexecerror",
         "//pkg/sql/gcjob",
         "//pkg/sql/isql",
         "//pkg/sql/lexbase",

--- a/pkg/ccl/testccl/sqlccl/explain_test.go
+++ b/pkg/ccl/testccl/sqlccl/explain_test.go
@@ -13,6 +13,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/internal/sqlsmith"
+	"github.com/cockroachdb/cockroach/pkg/sql/colexecerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/lexbase"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqltestutils"
 	"github.com/cockroachdb/cockroach/pkg/sql/tests"
@@ -136,6 +137,11 @@ func TestExplainGist(t *testing.T) {
 
 	skip.UnderDeadlock(t, "the test is too slow")
 	skip.UnderRace(t, "the test is too slow")
+
+	// Use the release-build panic-catching behavior instead of the
+	// crdb_test-build behavior. This is needed so that some known bugs like
+	// #117101 don't result in a test failure.
+	defer colexecerror.ProductionBehaviorForTests()()
 
 	ctx := context.Background()
 	rng, _ := randutil.NewTestRand()

